v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/transforms/AssignShardIdFn.java

```diff
@@ -0,0 +1,497 @@
+        } else {
+          String keysJsonStr = record.getMods().get(0).getKeysJson();
+          JsonNode keysJson = mapper.readTree(keysJsonStr);
+
+          String newValueJsonStr = record.getMods().get(0).getNewValuesJson();
+          JsonNode newValueJson = mapper.readTree(newValueJsonStr);
+          Map<String, Object> spannerRecord = new HashMap<>();
+          // Query the spanner database in case of a DELETE event
+          if (record.getModType() == ModType.DELETE) {
+            spannerRecord =
+                fetchSpannerRecord(
+                    record.getTableName(),
+                    record.getCommitTimestamp(),
+                    record.getServerTransactionId(),
+                    keysJson);
+          } else {
+            Table table = ddl.table(record.getTableName());
+
+            // Add all fields from keysJson and valuesJson to spannerRecord
+            for (Iterator<String> it = keysJson.fieldNames(); it.hasNext(); ) {
+              String key = it.next();
+              Column column = table.column(key);
+              spannerRecord.put(key, getColumnValueFromJson(column, keysJson));
+            }
+            for (Iterator<String> it = newValueJson.fieldNames(); it.hasNext(); ) {
+              String key = it.next();
+              Column column = table.column(key);
+              spannerRecord.put(key, getColumnValueFromJson(column, newValueJson));
+            }
+          }
+          ShardIdRequest shardIdRequest = new ShardIdRequest(record.getTableName(), spannerRecord);
+
+          ShardIdResponse shardIdResponse;
+          if (!customJarPath.isEmpty() && !shardingCustomClassName.isEmpty()) {
+            Distribution getShardIdResponseTimeMetric =
+                Metrics.distribution(AssignShardIdFn.class, "custom_shard_id_impl_latency_ms");
+            Instant startTimestamp = Instant.now();
+            shardIdResponse = shardIdFetcher.getShardId(shardIdRequest);
+            Instant endTimestamp = Instant.now();
+            getShardIdResponseTimeMetric.update(
+                new Duration(startTimestamp, endTimestamp).getMillis());
+          } else {
+            shardIdResponse = shardIdFetcher.getShardId(shardIdRequest);
+          }
+
+          qualifiedShard = shardIdResponse.getLogicalShardId();
+          if (qualifiedShard == null || qualifiedShard.isEmpty() || qualifiedShard.contains("/")) {
+            throw new IllegalArgumentException(
+                "Invalid logical shard id value: "
+                    + qualifiedShard
+                    + " for spanner table: "
+                    + record.getTableName());
+          }
+        }
+      }
+
+      record.setShard(qualifiedShard);
+      String tableName = record.getTableName();
+      String keysJsonStr = record.getMods().get(0).getKeysJson();
+      String finalKeyString = tableName + "_" + keysJsonStr + "_" + qualifiedShard;
+      Long finalKey = finalKeyString.hashCode() % maxConnectionsPerShard;
```

Comment:
Max connection per shard is a TODO added in the main class, this will eventually come from the shard file. I will handle this in a follow-up PR.For now, the code assumes all shards allow same max connections.